syncer/pipe.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"sync"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"github.com/siddontang/go/sync2"
+)
+
+// Pipe is the littlest process unit in syncer
+type Pipe interface {
+	// Name is the pipe's name
+	Name() string
+
+	// Init initializes the pipe
+	Init(*config.SubTaskConfig, func() error) error
+
+	// Update updates the pipe's config
+	Update(*config.SubTaskConfig)
+
+	// Input receives data
+	Input() chan *PipeData
+
+	// Output exports data already be processed
+	Output() chan *PipeData
+
+	// Run starts the pipe's process
+	Run()
+
+	// Close closes the pipe
+	Close()
+
+	// Wait waits all the data in the pipe is processed
+	Wait()
+
+	// SetErrorChan set the error channel, send error to this channel when meet error
+	SetErrorChan(chan error)
+
+	// SetResolveFunc set the resolveFunc, which used when data is resolved, used in the last pipe, or data is skiped
+	SetResolveFunc(func())
```

Comment:
can we do this from the `PipeData` of `Output`?

Reply:
because some data will be filter in pipe, so can't get from output. Or we add a variable in `PipeData` to let pipeline know this data should be filter?